libyul/interpreter/PureInterpreter.cpp

```diff
@@ -0,0 +1,378 @@
+}
+
+ExecutionResult PureInterpreter::operator()(Break const&)
+{
+	return ExecutionOk{ControlFlowState::Break};
+}
+
+ExecutionResult PureInterpreter::operator()(Continue const&)
+{
+	return ExecutionOk{ControlFlowState::Continue};
+}
+
+ExecutionResult PureInterpreter::operator()(Leave const&)
+{
+	return ExecutionOk{ControlFlowState::Leave};
+}
+
+ExecutionResult PureInterpreter::operator()(Block const& _block)
+{
+	enterScope(_block);
+	ScopeGuard guard([this] { leaveScope(); });
+
+	return execute(_block.statements);
+}
+
+ExecutionResult PureInterpreter::execute(std::vector<Statement> const& _statements)
+{
+	for (auto const& statement: _statements)
+	{
+		ExecutionResult statementRes = visit(statement);
+		if (statementRes != ExecutionResult(ExecutionOk{ControlFlowState::Default}))
+			return statementRes;
+	}
+	return ExecutionOk{ControlFlowState::Default};
+}
+
+ExecutionResult PureInterpreter::visit(Statement const& _statement)
+{
+	if (auto terminated = incrementStatementStep())
+		return *terminated;
+	return std::visit(*this, _statement);
+}
+
+EvaluationResult PureInterpreter::operator()(Literal const& _literal)
+{
+	return EvaluationOk(_literal.value.value());
+}
+
+EvaluationResult PureInterpreter::operator()(Identifier const& _identifier)
+{
+	auto it = m_variables.find(_identifier.name);
+	solAssert(it != m_variables.end());
+	return EvaluationOk(it->second);
+}
+
+EvaluationResult PureInterpreter::operator()(FunctionCall const& _functionCall)
+{
+	std::vector<std::optional<LiteralKind>> const* literalArguments = nullptr;
+	if (BuiltinFunction const* builtin = m_dialect.builtin(_functionCall.functionName.name))
+		if (!builtin->literalArguments.empty())
+			literalArguments = &builtin->literalArguments;
+	EvaluationResult argsRes = evaluateArgs(_functionCall.arguments, literalArguments);
+	if (auto* terminated = std::get_if<ExecutionTerminated>(&argsRes))
+		return *terminated;
+
+	std::vector<u256> argsValues = std::move(std::get<EvaluationOk>(argsRes).values);
+
+	if (EVMDialect const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
+	{
+		if (BuiltinFunctionForEVM const* builtinFunction = dialect->builtin(_functionCall.functionName.name))
+		{
+			PureEVMInstructionInterpreter interpreter(dialect->evmVersion());
+			return interpreter.evalBuiltin(*builtinFunction, _functionCall.arguments, argsValues);
+		}
+	}
+
+	FunctionDefinition const& functionDefinition = m_scope->getFunction(_functionCall.functionName.name);
+
+	yulAssert(argsValues.size() == functionDefinition.parameters.size());
+	VariableValuesMap variables;
+	for (size_t i = 0; i < functionDefinition.parameters.size(); ++i)
+		variables[functionDefinition.parameters.at(i).name] = argsValues.at(i);
+	for (size_t i = 0; i < functionDefinition.returnVariables.size(); ++i)
+		variables[functionDefinition.returnVariables.at(i).name] = 0;
+
+	std::unique_ptr<PureInterpreter> interpreter = makeInterpreterCopy(std::move(variables));
```

Comment:
This looks like it lives only until the end of the function. Why even allocate it on the heap?

Reply:
The heap allocation was copied from the original interpreter. Originally I did not think much about it, but then I think making heap allocation should help save more stack space, so we can allow more recursion calls.

But I also added the config to limit the number of recursion calls, so if you think this part does not have much effect on the stack, then I can change it to allocation on stack.

Reply:
Yeah, that was my impression too, but then if stack space was a real concern then it's only a half-measure and a proper solution would be to use a non-recursive implementation instead. I think we should just put it on stack and revisit if it turns out to be an actual problem. We should have a test uses nested calls up to the limit to verify that.

In practice we won't have much nesting anyway, so we could live even with a very low limit.

You'll probably find more questionable decisions like this in the old implementation. If it does not seem justified, we want it simplified. The general idea here is to treat it as a new implementation that's carefully reviewed from scratch.

Reply:
This is fixed.